### PR TITLE
gosnmp: add bounds checking for reqID & msgID

### LINKED ISSUE #272: fix overflows
Fix overflows in ``gosnmp.go`` and ``marshal.go`` where reqID or msgID could become > ``math.MaxInt32``

[RFC-3412#section-6](https://tools.ietf.org/html/rfc3412#section-6) mentions: ``msgID      INTEGER (0..2147483647),`` which is also to be used for ``requestID``

sub issue of #271 to fix ``TODO`` items in code base.

### PR DESCRIPTION
**Pull request by TimRots**
add bounds checking for reqID & msgID to prevent overflows as per [RFC-3412#section-6](https://tools.ietf.org/html/rfc3412#section-6):
``msgID      INTEGER (0..2147483647),``

Fixes #272

Signed-off-by: Tim Rots <tim.rots@protonmail.ch>

### CODE REVIEW

**Comment by SuperQ** [comment]
Looking at [rfc3412#section-6.2](https://tools.ietf.org/html/rfc3412#section-6.2), I don't see anything that forbids wrapping the counter. The msgID values are only supposed to be not re-used within a request.

**Comment by SuperQ** [comment]
Maybe @richih has a better understanding of this?

**Comment by brian-brazil** [comment]
> The msgID values are only supposed to be not re-used within a request.

Per 6.2 "Values for msgID SHOULD be generated in a manner that avoids re-use of any outstanding values", so wrapping around sounds fine to me. I doubt we're ever going to have 2 billion requests in flight.

**Comment by TimRots** [comment]
Thank you both for the review. Will leave things as-is then. 

Should we however get the TODO note removed?

**Comment by SuperQ** [comment]
We can't leave it as-is, because it's adding a to a uint32, not an int32. And we need to keep the IDs positive.

File: gosnmp.go [hunk 1]
```diff
@@ -429,11 +429,19 @@ func (x *GoSNMP) SnmpEncodePacket(pdutype PDUType, pdus []SnmpPDU, nonRepeaters
...
+	}
+
 	pkt.RequestID = reqID
 
 	if x.Version == Version3 {
-		msgID := atomic.AddUint32(&(x.msgID), 1) // TODO: fix overflows
+		msgID := atomic.AddUint32(&(x.msgID), 1)
+		if msgID > math.MaxInt32 {
+			return []byte{}, fmt.Errorf("msgID out of range: %v", reqID)
+		}
```
Comment (SuperQ) [inline gosnmp.go:443]:
```suggestion
		if msgID > math.MaxInt32 {
			msgID = 1
		}
```

File: gosnmp.go [hunk 2]
```diff
@@ -429,11 +429,19 @@ func (x *GoSNMP) SnmpEncodePacket(pdutype PDUType, pdus []SnmpPDU, nonRepeaters
 	pkt := x.mkSnmpPacket(pdutype, pdus, nonRepeaters, maxRepetitions)
 
 	// Request ID is an atomic counter (started at a random value)
-	reqID := atomic.AddUint32(&(x.requestID), 1) // TODO: fix overflows
+	reqID := atomic.AddUint32(&(x.requestID), 1)
+	if reqID > math.MaxInt32 {
+		return []byte{}, fmt.Errorf("reqID out of range: %v", reqID)
+	}
```
Comment (SuperQ) [inline gosnmp.go:435]:
```suggestion
	if reqID > math.MaxInt32 {
		reqID = 1
	}
```

**Comment by brian-brazil** [comment]
That doesn't sound right, whether it's an int32 or uint32 we need to avoid the highest bit being 1. So what we actually need is to mask that bit.

**Comment by TimRots** [comment]
Thanks. Will attempt to implement the suggestion Brian mentions while keeping in mind both your comments.

**Comment by TimRots** [comment]
Updated the PR so we wrap around  ``requestID`` and ``msgID`` if value > ``math.MaxInt32``.

```
2147483644 1111111111111111111111111111100
2147483645 1111111111111111111111111111101
2147483646 1111111111111111111111111111110
2147483647 1111111111111111111111111111111
0 0
1 1
2 10
3 11
```

Would this be the correct approach?

File: marshal.go [hunk 3]
```diff
@@ -173,14 +174,19 @@ func (x *GoSNMP) sendOneRequest(packetOut *SnmpPacket,
 			return nil, err
 		}
 
-		// Request ID is an atomic counter (started at a random value)
-		reqID := atomic.AddUint32(&(x.requestID), 1) // TODO: fix overflows
+		reqID := atomic.AddUint32(&(x.requestID), 1)
```
Comment (SuperQ) [inline marshal.go:177]: Should we preserve the comment here? It does seem a little out of place, as there's nothing about randomization here.

Comment (TimRots) [reply marshal.go:177]: should I just add ``// Request ID is an atomic counter`` instead of removing the entire comment?

Comment (SuperQ) [reply marshal.go:177]: How about this: S
```suggestion
		// Request ID is an atomic counter that wraps to 0 at max int32.
		reqID := atomic.AddUint32(&(x.requestID), 1)
```

Comment (TimRots) [reply marshal.go:177]: That looks better indeed :), will update the PR accordingly

Comment (TimRots) [reply marshal.go:178]: Done

**Review by SuperQ — APPROVED** [review]
LGTM

**Comment by SuperQ** [comment]
@brian-brazil Is this what you were thinking?

File: marshal.go [hunk 4]
```diff
@@ -173,14 +174,20 @@ func (x *GoSNMP) sendOneRequest(packetOut *SnmpPacket,
 			return nil, err
 		}
 
-		// Request ID is an atomic counter (started at a random value)
-		reqID := atomic.AddUint32(&(x.requestID), 1) // TODO: fix overflows
+		// Request ID is an atomic counter that wraps to 0 at max int32.
+		reqID := atomic.AddUint32(&(x.requestID), 1)
+		if reqID > math.MaxInt32 {
```
Comment (brian-brazil) [inline marshal.go:179]: This if is redundant, you can always apply the mask.

**Comment by brian-brazil** [comment]
This looks about right, but we can simplify the code a little.

**Comment by TimRots** [comment]
Thanks for another round of review gents. 
Will also remove the redundant if statements.

**Comment by brian-brazil** [comment]
:+1:

**Comment by SuperQ** [comment]
Yea, that's nicer.

**Pull request merged**